Revise document for handling exceptions for Two-phase Commit Transactions (#934)
scalar-labs/scalardb: docs/two-phase-commit-transactions.md

This document briefly explains how to execute Two-phase Commit Transactions in ScalarDB.

## Overview

> ScalarDB transactions normally execute in a single transaction manager instance.

Author: I applied the suggestion in 0db8458fcbd8fe14999231b2da301bff4fd0602b.

> Also, it might be helpful to have a diagram like this.
> https://speakerdeck.com/scalar/scalar-db-universal-transaction-manager?slide=27

Regarding this, let's do in another PR.

In that case, you begin a transaction, execute CRUD operations, and commit the transaction in the same transaction manager instance.

> In addition to normal transactions, ScalarDB also supports two-phase-commit-style transactions called *Two-phase Commit Transactions*.

Reviewer: Sorry, I may have capitalized `Two-phase Commit Transactions` since I thought it was referring to the name of the feature in ScalarDB. If this is a generic mention of `two-phase commit transactions`, let's lowercase it. 

And I would keep `normal transactions` lowercase because, as far as I know, it's not the name of a feature in ScalarDB.

Two-phase Commit Transactions execute a transaction that spans multiple transaction manager instances.
The transaction manager instances can be in the same process/applications or in different processes/applications.
For example, if you have transaction manager instances in multiple microservices, you can execute a transaction that spans multiple microservices.

In Two-phase Commit Transactions, there are two roles, a coordinator and a participant, that collaboratively execute a single transaction.

> A coordinator process and participant processes has another transaction manager instance.

Reviewer 2: ```suggestion
A coordinator process and participant processes all have different transaction manager instances.
```

The coordinator process first begins a transaction, and the participant processes join the transaction.

> After executing CRUD operations, the coordinator process and the participant processes execute the two-phase commit protocol to commit the transaction.

Reviewer 3: ```suggestion
After executing CRUD operations, the coordinator process and the participant processes commit the transaction by using the two-phase interface.
```
`the two-phase commit protocol` is not really defined here, so I felt it was difficult to parse.
I also thought it might be misleading since the protocols are both two-phase commits, but the interfaces are different in normal and two-phase commit transactions. so I suggested as above.

    factory1.getTwoPhaseCommitTransactionManager();

TransactionFactory factory2 =
    TransactionFactory.create("<configuration file path for the transaction manager2>");
TwoPhaseCommitTransactionManager transactionManager2 =
    factory2.getTwoPhaseCommitTransactionManager();

TwoPhaseCommitTransaction transaction1 = null;
TwoPhaseCommitTransaction transaction2 = null;
try {
  // Begin a transaction
  transaction1 = transactionManager1.begin();

  // Join the transaction begun by transactionManager1 with the transaction ID
  transaction2 = transactionManager2.join(transaction1.getId());

  // Execute CRUD operations in the transaction
  Optional<Result> result = transaction1.get(...);
  List<Result> results = transaction2.scan(...);
  transaction1.put(...);
  transaction2.delete(...);

  // Prepare the transaction
  transaction1.prepare();
  transaction2.prepare();

  // Validate the transaction
  transaction1.validate();
  transaction2.validate();

  // Commit the transaction. If any of the transactions succeeds to commit, you can regard the
  // transaction as committed
  AtomicReference<Exception> exception = new AtomicReference<>();
  boolean anyMatch =

>       Stream.of(transaction1, transaction2)

Reviewer 4: With this error handling, I think the commits can be executed in parallel as follows

```java
  Stream.of(transaction1, transaction2)
    .parallel()
    .anyMatch(
        ...
```

Reviewer 4: Ah, sorry. I found the parallel version code in the next section below.

This code is the sequential version, right? If so, please ignore my comment. But it might be good to mention sequential execution or something here?

Author: I think that has already been mentioned here:
https://github.com/scalar-labs/scalardb/pull/934/files#diff-2677f880f1a9e512f31cce8e12950863ddffd4be49408d310d72d4d73e6122b9R306

What do you think?

Reviewer 4: Looks good!

          .anyMatch(
              t -> {
                try {
                  t.commit();
                  return true;
                } catch (Exception e) {
                  exception.set(e);
                  return false;
                }
              });

  // If all the transactions fail to commit, throw the exception and rollback the transaction
  if (!anyMatch) {
    throw exception.get();
  }
} catch (Exception e) {
  // Rollback the transaction
  if (transaction1 != null) {

>     transaction1.rollback();

Reviewer 4: Wrapping this with try-catch would be safer just in case if `rollback()` could throw any exception.

Author: Fixed in 635b41d1ec91b4be3e98e835bd692a17a29407cb. Thanks.

            .map(
                t -> {
                  try {
                    t.commit();
                    return null;
                  } catch (TransactionException e) {
                    return e;
                  }
                })
            .filter(Objects::nonNull)
            .collect(Collectors.toList());

    // If any of the transactions succeeded to commit, you can regard the transaction as committed
    if (exceptions.size() < transactions.length) {
      if (!exceptions.isEmpty()) {
        // You can log the exceptions here if you want
      }

      return; // Succeeded to commit
    }

    //
    // If all the transactions failed to commit:
    //

    // If any of the transactions failed to commit due to `UnknownTransactionStatusException`, throw
    // it because you should not retry the transaction in such a case
    Optional<TransactionException> unknownTransactionStatusException =
        exceptions.stream().filter(e -> e instanceof UnknownTransactionStatusException).findFirst();
    if (unknownTransactionStatusException.isPresent()) {
      throw unknownTransactionStatusException.get();
    }

    // Otherwise, throw the first exception

>     throw exceptions.get(0);

Reviewer 4: [minor] This is a very corner case, but this line would throw NPE if `transactions` is empty?

Author: Ah, that's a very unexpected usage. We could add validation or an assertion to this method, but I think it might be too verbose since it's a sample code. What do you think?

Reviewer 4: Yeah, I think it's okay to go with current simple one 👍

If you catch `CommitConflictException`, it indicates that committing the transaction has failed due to transient faults (e.g., a conflict error). You can retry the transaction from the beginning.
If you catch `UnknownTransactionStatusException`, it indicates that the status of the transaction, whether it has succeeded or not, is unknown.
In such a case, you need to check if the transaction is committed successfully or not and retry it if it fails.
How to identify a transaction status is delegated to users.
You may want to create a transaction status table and update it transactionally with other application data so that you can get the status of a transaction from the status table.

Although not illustrated in the sample code, the `resume()` API could also throw a `TransactionNotFoundException`.
This exception indicates that the transaction associated with the specified ID was not found, and it might have been expired.
In such cases, you can retry the transaction from the beginning since the cause of this exception is basically transient.

> In the sample code, for `UnknownTransactionStatusException`, the transaction doesn't retry because the cause of the exception is nontransient.

Reviewer 4: > because the cause of the exception is nontransient

Maybe we should also mention that the transaction status is unknown (due to underlying database issue or something) and whether to retry must be decided by the application?

Author: I think that's mentioned here:
https://github.com/scalar-labs/scalardb/pull/934/files#diff-2677f880f1a9e512f31cce8e12950863ddffd4be49408d310d72d4d73e6122b9R676-R679

What do you think?

Reviewer 4: Yeah, the previous description looks perfect.

But, I have a concern that the readers might forget the previous important explanation about UnknownTransactionStatusException since this line only mentions nontrasient causes.

In my understanding, UnknownTransactionStatusException must not be retried, while other non-transient exceptions can be retried although it may make no sense. Just mentioning nontransient causes seems a bit weak in preventing users from attempting retries.

Author: That makes sense.

So how about this:
```suggestion
In the sample code, for `UnknownTransactionStatusException`, the transaction doesn't retry as mentioned earlier.
```

Reviewer 4: Looks good!

As you can see, by resuming the transaction, you can share the same transaction object across multiple endpoints in `ServiceB`.

### Handle exceptions

In the previous section, you saw [how to execute a transaction with multiple transaction manager instances](#execute-a-transaction-with-multiple-transaction-manager-instances). However, you may also need to handle exceptions properly.
This section describes how to handle exceptions in Two-phase Commit Transactions.

Two-phase commit transactions are basically executed by multiple processes/applications (a coordinator and participants).
However, in this example code, you can see multiple transaction managers (`transactionManager1` and `transactionManager2`) in a single process.

> The following example code shows how to handle exceptions in Two-phase Commit Transactions:

Author: Fixed in 272d88b46c4f21fc6da8001800596115f754d637. Can you please check it?

Requests in the same gRPC connection always go to the same server in L3/L4 load balancing.
When using an L7 load balancer, since requests in the same gRPC connection don't necessarily go to the same server, you need to use cookies or similar method to route requests to the correct server.
For example, if you use [Envoy](https://www.envoyproxy.io/), you can use session affinity (sticky session) for gRPC.
Alternatively, you can use [bidirectional streaming RPC in gRPC](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc) since the L7 load balancer distributes requests in the same stream to the same server.

### HTTP/1.1

Typically, you use a server-side (proxy) load balancer with HTTP/1.1.
When using an L3/L4 load balancer, you can use the same HTTP connection to send requests in a transaction, which guarantees the requests go to the same server.
When using an L7 load balancer, since requests in the same HTTP connection don't necessarily go to the same server, you need to use cookies or similar method to route requests to the correct server.
You can use session affinity (sticky session) in that case.

> 

Author: Yes, we should. Let's do that in another PR. Thanks.

    throw exceptions.get(0);
  }

  private static void rollback(TwoPhaseCommitTransaction... transactions) {
    Stream.of(transactions)
        .parallel()
        .filter(Objects::nonNull)
        .forEach(
            t -> {
              try {
                t.rollback();
              } catch (RollbackException e) {
                // Rolling back the transaction failed. As the transaction should eventually
                // recover, you don't need to do anything further. You can simply log the occurrence
                // here
              }
            });
  }
}
```

The `begin()` API could throw `TransactionException` and `TransactionNotFoundException`.
If you catch `TransactionException`, it indicates that the transaction has failed to begin due to transient or nontransient faults. You can try retrying the transaction, but you may not be able to begin the transaction due to nontransient faults.
If you catch `TransactionNotFoundException`, it indicates that the transaction has failed to begin due to transient faults. You can retry the transaction.

The `join()` API could also throw a `TransactionException` and `TransactionNotFoundException`.

> And the way to handle them is the same as the `begin()` API.

Reviewer: ```suggestion
You can handle these exceptions in the same way that you handle the exceptions for the `begin()` API.
```

In such a case, you will exhaust the number of retries.

Please note that if you begin a transaction by specifying a transaction ID, you must use a different ID when you retry the transaction.
And, in the sample code, the transaction retries three times maximum and sleeps for 100 milliseconds before it retries.
But you can choose a retry policy, such as exponential backoff, according to your application requirements.

## Request Routing in Two-phase Commit Transactions

Services using Two-phase Commit Transactions usually execute a transaction by exchanging multiple requests and responses as follows:

![](images/two_phase_commit_sequence_diagram.png)

Also, each service typically has multiple servers (or hosts) for scalability and availability and uses server-side (proxy) or client-side load balancing to distribute requests to the servers.
In such a case, since a transaction processing in Two-phase Commit Transactions is stateful, requests in a transaction must be routed to the same servers while different transactions need to be distributed to balance the load.

![](images/two_phase_commit_load_balancing.png)

> There are several approaches to achieve it depending on the protocol between the services. Here, we introduce some approaches for gRPC and HTTP/1.1.

Author: `to achieve it` means `requests in a transaction must be routed to the same servers`.
